python/runtime/cudaq/algorithms/py_run.cpp

```diff
@@ -0,0 +1,168 @@
+              return readPyObject<double>(ty, results.data + i);
+            })
+            .Case([&](Float32Type ty) -> py::object {
+              return readPyObject<float>(ty, results.data + i);
+            })
+            .Default([](Type ty) -> py::object {
+              ty.dump();
+              throw std::runtime_error("Invalid return type for cudaq.run.");
+            });
+    ret.push_back(obj);
+  }
+  return ret;
+}
+
+std::tuple<std::string, MlirModule, OpaqueArguments *,
+           mlir::ArrayRef<mlir::Type>>
+getKernelLaunchParameters(py::object &kernel, py::args args) {
+  if (py::len(kernel.attr("arguments")) != args.size())
+    throw std::runtime_error("Invalid number of arguments passed to run:" +
+                             std::to_string(args.size()) + " expected " +
+                             std::to_string(py::len(kernel.attr("arguments"))));
+
+  if (py::hasattr(kernel, "compile"))
+    kernel.attr("compile")();
+
+  auto kernelName = kernel.attr("name").cast<std::string>();
+  auto kernelMod = kernel.attr("module").cast<MlirModule>();
+  args = simplifiedValidateInputArguments(args);
+  auto *argData = toOpaqueArgs(args, kernelMod, kernelName);
+
+  auto returnTypes = getKernelFuncOp(kernelMod, kernelName).getResultTypes();
+  return {kernelName, kernelMod, argData, returnTypes};
+}
+} // namespace
+
+/// @brief Run `cudaq::run` on the provided kernel.
+std::vector<py::object> pyRun(py::object &kernel, py::args args,
+                              py::int_ shots_count,
+                              std::optional<noise_model> noise_model) {
+  auto [kernelName, kernelMod, argData, returnTypes] =
+      getKernelLaunchParameters(kernel, args);
+
+  if (returnTypes.empty() || returnTypes.size() > 1)
+    throw std::runtime_error(
+        "cudaq.run only supports kernels that return a value.");
+
+  if (shots_count < 0)
+    throw std::runtime_error("Invalid shots_count. Must be non-negative.");
+
+  auto returnTy = returnTypes[0];
+  auto mod = unwrap(kernelMod);
+  mod->setAttr(cudaq::runtime::enableCudaqRun,
+               mlir::UnitAttr::get(mod->getContext()));
+
+  auto &platform = cudaq::get_platform();
+  if (noise_model.has_value()) {
+    if (platform.is_remote())
+      throw std::runtime_error(
+          "Noise model is not supported on remote platforms.");
+  }
+
+  if (shots_count == 0)
+    return {};
+
+  // Launch the kernel in the appropriate context.
+  if (noise_model.has_value())
+    platform.set_noise(&noise_model.value());
+
+  details::RunResultSpan results = details::runTheKernel(
+      [&]() mutable { pyAltLaunchKernel(kernelName, kernelMod, *argData, {}); },
+      platform, kernelName, shots_count);
+  delete argData;
+
+  if (noise_model.has_value())
+    platform.reset_noise();
+
+  mod->removeAttr(cudaq::runtime::enableCudaqRun);
+  return readRunResults(returnTy, results);
+}
+
+/// @brief Bind the run cudaq function
+void bindPyRun(py::module &mod) {
```

Comment:
```suggestion
void cudaq::bindPyRun(py::module &mod) {
```

Reply:
this failed to compile with python bindings - it wants me to declare bindPyRun in cudaq namespace